### PR TITLE
Support defaulting to namespace placement for namespace-scoped federation

### LINKED ISSUE #298: Ability to create or remove all resources in namespace with namespace-scoped deployment
With a cluster-scoped federation control plane deployment, we currently support creation and removal of all resources in a namespace by creating or deleting the entire namespace in target clusters via `FederatedNamespacePlacement`. With a namespace-scoped deployment, the existence of a namespace is not up to the federation controller as we no longer have a `Namespace` sync controller (after #297 is fixed), nor does the federation controller have permissions to create/delete namespaces.

Even with those limitations, we need the ability to create or remove all resources in a namespace for namespace-scoped deployments as we have for cluster-scoped. The fix for #75 would be part of this.

### PR DESCRIPTION
**Pull request by marun**
In cluster-scoped federation, the namespace sync controller uses namespace placement to determine whether a namespace should appear in member clusters. If a namespace does not appear in member clusters, any resources it contains are also prevented from appearing.

Since namespace-scoped federation does not run the namespace sync controller, namespace placement would have no effect.  This PR changes that by ensuring that, for namespaced federation, placement for a given resource is the determined from the resource and namespace placement.

Fixes #298

### CODE REVIEW

File: test/e2e/placement.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			}
+			err = wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (bool, error) {
+				_, err := client.Resources(namespace).Get(qualifiedName.Name, metav1.GetOptions{})
+				if errors.IsNotFound(err) {
+					return true, nil
+				}
+				if err != nil {
+					tl.Errorf("Failed to check for existence of %s %q in cluster %q: %v",
+						targetKind, qualifiedName, clusterName, err,
+					)
```
[the remark this answers is not in the document]
Comment (marun) [reply test/e2e/placement.go]: Is that a problem?  I like this style of multi-line function call.

File: pkg/controller/sync/controller.go [hunk 2]
```diff
@@ -176,8 +176,22 @@ func newFederationSyncController(typeConfig typeconfig.Interface, kubeConfig *re
 	targetAPIResource := typeConfig.GetTarget()
 	if targetAPIResource.Kind == util.NamespaceKind {
 		s.placementPlugin = placement.NewNamespacePlacementPlugin(placementClient, targetNamespace, enqueueObj)
-	} else {
+	} else if targetNamespace == metav1.NamespaceAll {
 		s.placementPlugin = placement.NewResourcePlacementPlugin(placementClient, targetNamespace, enqueueObj)
+	} else {
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/controller.go]: What do you think would be the advantage(s) of using `NamespacedPlacementPlugin` regardless of whether the controller was namespace or cluster-scoped?  The cost of namespaced placement is an extra informer for `FederatedNamespacePlacement` per sync controller.

File: test/e2e/placement.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (font) [reply test/e2e/placement.go]: I'm all for multi-line function calls. It's just inconsistent to have the closing `)` on its own line when none of the other function calls do this. But we don't necessarily have a style guide so who am I to enforce it?

File: pkg/controller/sync/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (font) [reply pkg/controller/sync/controller.go]: Yes, better error handling would do the trick too. In regards to getting the `namespace does not exist` error, I'm wondering if error handling would require looking at namespace placement to determine if the namespace should really exist, or if it was somehow removed for another reason in which case the error should really be logged.

**Comment by marun** [comment]
@font Regarding having multi-line function calls, this is gofmt-compliant.  I discovered it accidentally due to having gofmt enabled on-save.

**Comment by font** [comment]
> @font Regarding having multi-line function calls, this is gofmt-compliant. I discovered it accidentally due to having gofmt enabled on-save.

Yes, I figured it was. I think it's because there is a trailing `,` on the last variable.

**Pull request merged**